File: 프로그래머스/1/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    answer = []
    new_park = []
    N, M = len(park), len(park[0])
    
    for i in range(N):
        new_park.append(list(park[i]))
    
    # print(new_park)
    
    directions = {'N': (-1,0), 'S': (1,0), 'W': (0,-1), 'E': (0, 1)}
    
    def find_si_sj():
        for i in range(N):
            for j in range(M):
                if new_park[i][j]=='S':
                    return i, j
    
    ci, cj = find_si_sj() #현재위치
    
    for route in routes:
        dir, cnt = route.split(" ")
        di, dj = directions[dir]
        # print(di, dj)
        
        flag = True
        ni, nj = ci, cj
        for i in range(int(cnt)):
            ni+=di
            nj+=dj

            if not(0<=ni<N and 0<=nj<M and new_park[ni][nj]!='X'):
                flag = False
                break
        if flag:
            ci, cj = ni, nj
        
        # print(ci, cj)
    
    
    return [ci, cj]

File: 프로그래머스/1/172928. 공원 산책/공원 산책.java
import java.util.*;

class Solution {
    static boolean possible(int ni, int nj, int N, int M, int[][] arr) {
        // 범위 체크
        if (ni<0||ni>=N||nj<0||nj>=M) {
            return false;
        }
        // 장애물 체크
        if (arr[ni][nj]==1) {
            return false;
        }
        
        return true;
    }
    public int[] solution(String[] park, String[] routes) {
        int[] answer = {};
        int N = park.length;
        int M = park[0].length();
        int[][] arr = new int[N][M];
        int si=-1;
        int sj=-1;
        
        for (int i=0; i<N; i++) {
            for (int j=0; j<M; j++) {
                if (park[i].charAt(j)=='S') {
                    arr[i][j]=2;
                    si = i;
                    sj = j;
                }
                else if (park[i].charAt(j)=='O') {
                    arr[i][j]=0;
                }
                else {
                    arr[i][j]=1;
                }
            }
        }
        
        for (String route: routes) {
            String[] token = route.split(" ");
            int n = Integer.parseInt(token[1]);
            
            int ni = si;
            int nj = sj;
            
            boolean flag = true;
                        
            if (token[0].equals("N")) {
                int i = 0;
                while (i<n) {
                    ni-=1;
                    flag = possible(ni, nj, N, M, arr);
                    if (!flag) break;
                    i++;
                }
            }
            else if (token[0].equals("S")) {
                int i = 0;
                while (i<n) {
                    ni+=1;
                    flag = possible(ni, nj, N, M, arr);
                    if (!flag) break;
                    i++;
                }
            }
            if (token[0].equals("W")) {
                int i = 0;
                while (i<n) {
                    nj-=1;
                    flag = possible(ni, nj, N, M, arr);
                    if (!flag) break;
                    i++;
                }
            }
            if (token[0].equals("E")) {
                int i = 0;
                while (i<n) {
                    nj+=1;
                    flag = possible(ni, nj, N, M, arr);
                    if (!flag) break;
                    i++;
                }
            }
            
            // System.out.println(ni+" "+nj);
            
            if (flag) {
                si=ni;
                sj=nj;
            }
            
            // System.out.println(si+" "+sj);
        }
        
        return new int[]{si, sj};
    }
}